libraries/AP_RangeFinder/AP_RangeFinder_BLPing.cpp

```diff
@@ -0,0 +1,286 @@
+    uint16_t count = 0;
+
+    // read any available lines from the lidar
+    int16_t nbytes = uart->available();
+    while (nbytes-- > 0) {
+        int16_t b = uart->read();
+        if (b < 0) {
+            continue;
+        }
+        if (parse_byte(b)) {
+            count++;
+            sum_cm += distance_cm;
+        }
+    }
+
+    if (count > 0) {
+        // return average distance of readings
+        reading_cm = sum_cm / count;
+
+        // request another distance
+        send_message(BLDPIN_MSGID_DISTANCE_SIMPLE, nullptr, 0);
+
+        return true;
+    }
+
+    // no readings so return false
+    return false;
+}
+
+// process one byte received on serial port
+// returns true if a distance message has been successfully parsed
+// state is stored in msg structure
+bool AP_RangeFinder_BLPing::parse_byte(uint8_t b)
+{
+    bool got_distance = false;
+
+    // process byte depending upon current state
+    switch (msg.state) {
+
+    case MSG_STATE_HEADER1:
+        if (b == BLPING_FRAME_HEADER1) {
+            msg.crc_expected = BLPING_FRAME_HEADER1;
+            msg.state = MSG_STATE_HEADER2;
+        }
+        break;
+
+    case MSG_STATE_HEADER2:
+        if (b == BLPING_FRAME_HEADER2) {
+            msg.crc_expected += BLPING_FRAME_HEADER2;
+            msg.state = MSG_STATE_LEN_L;
+        } else {
+            msg.state = MSG_STATE_HEADER1;
+        }
+        break;
+
+    case MSG_STATE_LEN_L:
+        msg.payload_len = b;
+        msg.crc_expected += b;
+        msg.state = MSG_STATE_LEN_H;
+        break;
+
+    case MSG_STATE_LEN_H:
+        msg.payload_len += ((uint16_t)b << 8);
+        msg.payload_recv = 0;
+        msg.crc_expected += b;
+        msg.state = MSG_STATE_MSG_ID_L;
+        break;
+
+    case MSG_STATE_MSG_ID_L:
+        msg.msgid = b;
+        msg.crc_expected += b;
+        msg.state = MSG_STATE_MSG_ID_H;
+        break;
+
+    case MSG_STATE_MSG_ID_H:
+        msg.msgid += ((uint16_t)b << 8);
+        msg.crc_expected += b;
+        msg.state = MSG_STATE_SRC_ID;
+        break;
+
+    case MSG_STATE_SRC_ID:
+        msg.crc_expected += b;
+        msg.state = MSG_STATE_DST_ID;
+        break;
+
+    case MSG_STATE_DST_ID:
+        msg.crc_expected += b;
+        msg.state = MSG_STATE_PAYLOAD;
+        break;
+
+    case MSG_STATE_PAYLOAD:
+        if (msg.payload_recv < msg.payload_len) {
+            msg.payload[msg.payload_recv] = b;
+            msg.payload_recv++;
+            msg.crc_expected += b;
+        }
+        if (msg.payload_recv == msg.payload_len) {
+            msg.state = MSG_STATE_CRC_L;
+        }
+        break;
+
+    case MSG_STATE_CRC_L:
+        msg.crc = b;
+        msg.state = MSG_STATE_CRC_H;
+        break;
+
+    case MSG_STATE_CRC_H:
+        msg.crc += ((uint16_t)b << 8);
+        msg.state = MSG_STATE_HEADER1;
+        if (msg.crc_expected == msg.crc) {
+
+            // process payload
+            switch(msg.msgid) {
+
+            case BLPING_MSGID_ACK:
```

Comment:
This switch is a fair indication that this should be an `enum class`, not a list of defines.

Reply:
I tried to make it an enum but it because messy in the parser which could a message id that isn't listed in the enum and I'm not too keen on adding every message id that the sensor supports (there are a lot!)

Reply:
https://github.com/ArduPilot/ardupilot/pull/10938/files#diff-808bc25d8005172cc2ec3a3b7de2fac4R33 <- it's only those defines that matter, right?

Reply:
@rmackay9 You could use an enum and add a `default` handler, no need to handle every message (or every integer in the 16-bit space for that matter).

Reply:
You also only need to put the messages you are interested in using in the enum.